File: typescript/packages/core/src/commands/builtin/ram/du.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { du as ramDu, duAll as ramDuAll } from '../../../core/ram/du.ts'
import type { RAMAccessor } from '../../../accessor/ram.ts'
import { ResourceName } from '../../../types.ts'
import { command } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { duGeneric } from '../generic/du.ts'

export const RAM_DU = command({
  name: 'du',
  resource: ResourceName.RAM,
  spec: specOf('du'),
  fn: (accessor: RAMAccessor, paths, _texts, opts) =>
    duGeneric(
      paths,
      opts,
      (p) => ramDu(accessor, p),
      (p) => ramDuAll(accessor, p),
    ),
})

File: python/mirage/commands/builtin/nextcloud/du.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from functools import partial

from mirage.accessor.nextcloud import NextcloudAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.generic.du import du_multi
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.nextcloud.du import du as du_impl
from mirage.core.nextcloud.du import du_all as du_all_impl
from mirage.core.nextcloud.glob import resolve_glob
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


@command("du", resource="nextcloud", spec=SPECS["du"])
async def du(
    accessor: NextcloudAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: bytes | None = None,
    h: bool = False,
    s: bool = False,
    a: bool = False,
    max_depth: str | None = None,
    c: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    paths = await resolve_glob(accessor, paths, index)
    out = await du_multi(
        paths,
        compute_total=partial(du_impl, accessor),
        compute_all=partial(du_all_impl, accessor),
        h=h,
        s=s,
        a=a,
        max_depth=int(max_depth) if max_depth is not None else None,
        c=c,
    )
    return out, IOResult()
